domain/user/state/state.go

```diff
@@ -0,0 +1,120 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/user"
+	"github.com/juju/juju/domain"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// AddUser will add a new user to the database. If the user already exists
+// an error that satisfies usererrors.AlreadyExists will be returned. If the
+// users creator is set and does not exist then an error that satisfies
+// usererrors.UserCreatorNotFound will be returned.
+func (st *State) AddUser(ctx context.Context, user user.User) error {
+	db, err := st.DB()
+	if err != nil {
+		return fmt.Errorf("cannot get DB access")
+	}
+
+	userUUID, err := utils.NewUUID()
+	if err != nil {
+		return fmt.Errorf("cannot generate UUID for user %q: %v", user.Name, err)
+	}
+	q := "INSERT INTO user (uuid, name, display_name, removed, created_by_uuid, created_at) VALUES (?, ?, ?, ?, ?, ?)"
+
+	if user.CreatorUUID == "" {
+		// this is Admin user, that why they are self-created
+		user.CreatorUUID = userUUID.String()
+	}
+
+	err = db.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		_, err := tx.ExecContext(ctx, q,
+			userUUID.String(),
+			user.Name,
+			user.DisplayName,
+			false,
+			user.CreatorUUID,
+			user.CreatedAt,
+		)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		return nil
+	})
+	if err != nil {
```

Comment:
If you take a look in `internal/database/errors.go` there is a set of helper functions for helping you fulfil the contract being offered by state.

Example:
```go
if database.IsErrConstraintUnique(err) { // AlreadyExists error }
if database.IsErrConstraintForeignKey(err) { // UserCreatorNotFound error }
```

Reply:
thx, I hope I used these utils correctly.